### ISSUE TITLE
`gr.update()` doesn't support updating parameters that affect the preprocessing/postprocessing in backend

### ISSUE THREAD

**Issue by NimaBoscarino** [gradio-app/gradio#1566]
### Describe the bug

When `gr.Radio.update(choices=choices)` is called on a radio component that has `type="index"`, the radio component breaks because the underlying `self.choices` (server-side) isn't updated. So when the call to `return self.choices.index(x)` [happens here](https://github.com/gradio-app/gradio/blob/d455baca4ed982846e17b5f27b5a9429b9944f1a/gradio/components.py#L1173), an error gets thrown.

Using `type="value"` isn't affected by this. (In general, I don't think any of the underlying `self.ETC` values are actually updated, but I haven't dug too deep into the code)

~~I intend on spending some time trying to find a fix for it!~~

>[ I think the solution here needs to be that we pass in the updated component configs from the frontend into the backend when we call the /api/predict endpoint.](https://github.com/gradio-app/gradio/pull/1568#issuecomment-1157008364) - @abidlabs 

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

https://colab.research.google.com/drive/1SgBexhut4rFFg0i6BPo_o9ZBcnmrZ2yH#scrollTo=0EGvW7wzCGfW

### Logs

```shell
File "/Users/nima/.pyenv/versions/3.10.4/envs/sandbox/lib/python3.10/site-packages/gradio/components.py", line 1173, in preprocess
    return self.choices.index(x)
ValueError: 'A' is not in list
```

### System Info

```shell
Gradio 3.0.17, MacOS 12.4.
```

### Severity

annoying

**Comment by abidlabs** [comment]
Thanks @NimaBoscarino for the detailed issue and for looking into a solution!

**Comment by NimaBoscarino** [comment]
Un-assigning myself since I don't think I'll have time work on this in the next few weeks, so it's up for grabs!

**Comment by abidlabs** [comment]
Finally closed via #5240!

**Comment by AndreiSedyshev** [comment]
This is still not fixed by #5240 

This change does not work:
https://github.com/gradio-app/gradio/blob/da05e59a53bbad15e5755a47f46685da18e1031e/demo/variable_outputs/run.py

**Comment by AndreiSedyshev** [comment]
> Hi @AndreiSedyshev you'll need to install from main, which you can do by following the instructions here: https://gradio.app/docs/main

Tremendous, I apologize I wasn't installing it correctly. Love this change, thank you so much

**Comment by Solomin0** [comment]
@abidlabs  @AndreiSedyshev  Did this ever get solved? I am on version 4.31.0 and it is not working. The visual dataset is updated but when clicking the original index it will produce the same initial unchanged value. 

```
import gradio as gr
def update(name):

    print(name)
    # new_examples = gr.Examples(["1","2","3"], inputs=[inp],label = "Saved Prompts")
    # return f"Welcome to Gradio, {name}!", new_examples
    return [["a"]]
with gr.Blocks() as demo:
    with gr.Row():
        inp = gr.Textbox(placeholder="What is your name?")

        saved_examples = gr.Dataset(components = ["text"], samples = [["This is my Favorite Prompt"], [""]],label= "Saved Prompts" )

        saved_examples.click(update, inputs=[saved_examples], outputs=[saved_examples])
    # btn.click(fn=update, inputs=inp, outputs=[ex])

demo.launch()
```
